src/module/actor/character/crafting/entry.ts

```diff
@@ -117,11 +114,13 @@ export class CraftingEntry implements CraftingEntryData {
             return false;
         }
 
-        if (!this.requiredTraits.some((traits) => traits.every((t) => formula.item.traits.has(t)))) {
+        const predicate = new PredicatePF2e(this.craftableItems ?? {});
+
+        if (!predicate.test(formula.item.getRollOptions("item"))) {
             if (warn) {
                 ui.notifications.warn(
                     game.i18n.format("PF2E.CraftingTab.Alerts.ItemMissingTraits", {
-                        traits: JSON.stringify(this.requiredTraits),
+                        traits: JSON.stringify(this.craftableItems),
                     })
```

Comment:
I wouldn't attempt to just dump predicate JSON into a warning message. Exclude it and just say the item is ineligible.